src/content/1.7/development/components/grid/tutorials/work-with-search-form/custom-form-factory.md

```diff
@@ -0,0 +1,146 @@
+---
+title: Create your custom form factory
+menuTitle: Custom form factory
+weight: 1
+---
+
+# Create your custom form factory
+
+In some cases the generic form factory is not enough for your need. In PrestaShop the category grid has two specific features:
+
+- it can manage a parent category which changes the grid rendering (only displays children) so it needs to dynamically add a parameter
+- its search url doesn't match the list so it needs to specify the form action
+- the redirection url is dynamic so that it takes the parent category into account 
+
+To manage this additional behaviour we create a decorator that uses the default form factory but adds specific behaviour.
+
+## Decorate the default factory
+
+{{% notice note %}}
+In those example we use decorator that are based on the default grid factory. But you could create a fully autonomous class as long as it implements the `GridFilterFormFactoryInterface`.
+{{% /notice %}}
+
+### Factory decorator
+
+```php
+/**
+ * Class CategoryFilterFormFactory decorates original filter factory to add custom submit action.
+ */
+final class CategoryFilterFormFactory implements GridFilterFormFactoryInterface
+{
+    /**
+     * @var GridFilterFormFactoryInterface
+     */
+    private $formFactory;
+
+    /**
+     * @var UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @param GridFilterFormFactoryInterface $formFactory
+     * @param UrlGeneratorInterface $urlGenerator
+     * @param RequestStack $requestStack
+     */
+    public function __construct(
+        GridFilterFormFactoryInterface $formFactory,
+        UrlGeneratorInterface $urlGenerator,
+        RequestStack $requestStack
+    ) {
+        $this->formFactory = $formFactory;
+        $this->urlGenerator = $urlGenerator;
+        $this->requestStack = $requestStack;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create(GridDefinitionInterface $definition)
```

Comment:
I'll remove the form changes and only leave the url changes

Reply:
Actually not that simple the form builder is recreated based on the generated form, because the setAction can't be called on the form itself..

Reply:
Looking at this in more details, there's actually few simple things you can do by decoration this form factory because you only have access to the GridDefinition, and the generated form:
- modifying the definition could be done here as a simple example, but it's not the correct way to do it since we already have the `*GridDefinitionModifier` that is made for that
- modifying the generated form is not possible, the `FormInterface` nearly only has getters, we could add some extra form fields, but here again it should be handled via modifying the definition filters

What would have been nice is to be able to get the form builder from the core factory, but the only function in `GridFilterFormFactoryInterface` is `create` and we can't modify it to add a `getFormBuilder` function (this would have made the action change much easier).

So in any was decorating this service will always be used for advanced modification, so having an advanced example makes sense.

Unless you have an idea of a simple use case, in which case we will replace this one, I'd say we keep this one for now.